Context menu to toggle error bars on plot

dtasev opened the pull request:
Will be wrapped up after ~~#25945~~ is merged

**Description of work.**
Adds context menu to toggle error bars on plot.

Errors for each line can be toggled individually, or all errors can be toggled/shown/hidden.

**To test:**
#### Setup
Script for **8** script plots and 2 workspaces.
```python
# The following line helps with future compatibility with Python 3
# print must now be used as a function, e.g print('Hello','World')
from __future__ import (absolute_import, division, print_function, unicode_literals)

# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *

import matplotlib.pyplot as plt

import numpy as np
ws2d_histo = CreateWorkspace(DataX=[10, 20, 30, 10, 20, 30, 10, 20, 30],
                             DataY=[2, 3, 4, 5, 3, 5],
                             DataE=[1, 2, 3, 4, 1, 1],
                             NSpec=3,
                             Distribution=True,
                             UnitX='Wavelength',
                             VerticalAxisUnit='DeltaE',
                             VerticalAxisValues=[4, 6, 8])

ws2d_1spec = CreateWorkspace(DataX=[10, 20, 30],
                                    DataY=[2, 3],
                                    DataE=[1, 2],
                                    NSpec=1,
                                    Distribution=True,
                                    UnitX='Wavelength',
                                    VerticalAxisUnit='DeltaE',
                                    VerticalAxisValues=[4])
                                    
def do_plots(bonus_kwargs=None):
    if bonus_kwargs is None:
        bonus_kwargs = {}
    plt.figure()
    plt.title('Script Plot without Errors')
    plt.plot([0,15000], [0,15000], label='MyLabel')
    plt.plot([0,15000], [0,14000], label='MyLabel 2')
    plt.legend()

    x = np.linspace(0, 15000, 150)
    plt.figure()
    plt.title('Script Plot with YErr and XErr.{}'.format('' if len(bonus_kwargs) == 0 else " Bonus ARGS: {}".format(bonus_kwargs)))
    plt.errorbar(x, 2*x, yerr=5*np.sqrt(2*x), xerr=2*np.sqrt(2*x), label='MyLabel', **bonus_kwargs)
    plt.legend()

    x = np.linspace(0, 15000, 150)
    plt.figure()
    plt.title('Script Plot with ONLY YErr.{}'.format('' if len(bonus_kwargs) == 0 else " Bonus ARGS: {}".format(bonus_kwargs)))
    plt.errorbar(x, 2*x, yerr=5*np.sqrt(2*x), label='MyLabel', **bonus_kwargs)
    plt.legend()

    x = np.linspace(0, 15000, 150)
    plt.figure()
    plt.title('Script Plot with ONLY XErr.{}'.format('' if len(bonus_kwargs) == 0 else " Bonus ARGS: {}".format(bonus_kwargs)))
    plt.errorbar(x, 2*x, xerr=2*np.sqrt(2*x), label='MyLabel', **bonus_kwargs)
    plt.legend()

do_plots()
do_plots({'capsize':4})
plt.show()
```

#### To Test
- `ws2d_1spec` - check that the error bar for the 1 spectra can be toggled from:
  - Right Click > Error Bars > Show/Hide all errors
  - Right Click > Error Bars > click on the 1 spectra shown
  - Cogwheel for plot options > Curves > Errorbars > check/uncheck `hide errorbars`
- `ws2d_histo`
  - Same as above, but for more than one curve
- Scripted plots
  - Check that Y and X errors are toggled correctly
    - Toggling the errors should change the state of both Y and X errors at the same time
  - Check that the one without errors doesn't provide the context menu at all
<!-- Instructions for testing. -->
- Check that plots can be saved and loaded correctly, before and after toggling the errorbars.

Fixes #25547

Release notes added in `docs/source/release/v4.1.0/ui.rst`.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

dtasev committed:
Can now toggle error bars on errorbar plot, re #25547

dtasev committed:
Will now add errors to a line, loses legend position

Re #25547

dtasev committed:
Legend now keeps the order for lines with errors

The `Axes.legend` function was overriden to provide a custom order for
`Axes.legend()` calls. The new logic is only executed if there are no
custom arguments passed to the function, and there is at least one line
with errors plotted.

Re #25547

dtasev committed:
Plots now save/load correctly, re #25547

Moved functions for managing errors into the FigureErrorsManager class.
Added `make_visible` argument to force state of error lines.
Added docstrings.

dtasev committed:
Merge branch 'master' into 25547_toggle_errorbars_on_plot

dtasev committed:
Added release notes, re #25547

dtasev committed:
Simplified adding logic, now it all goes through a single func

Re #25547

dtasev committed:
Added tests for changes, re #25547

dtasev committed:
Hide labels starting with _, re #25547

dtasev committed:
Remove unused import, re #25547

dtasev committed:
Do not try to retrieve legend for lines with no labels and no errors

Re #25547

DTasev commented:
Good spot! I've added the check. What should be done about scripted plots that  contain errors though? Should the menu still be shown?

DTasev commented:
Another thing I've left in by accident is the show/hide all buttons in the context menu. My main reason for adding those was to test the functions as they will likely be used for the plot settings dialog's `Apply to All`. 

Should I remove those from the context menu?

dtasev committed:
Added tests for interaction with scripted plots

- Moved `find_errorbar_container` function to plots.utility

- Added a check for MantidAxes and errorbars when the context menu is
requested. Menu actions are added only:
  - if the axes is a MantidAxes
  - if the axes has at least one line with errors, then the context menu
shows an entry only for that one line

- The _toggle all function will raise a NotImplementedError if called
with a line that is not on a MantidAxes. This error is for developers
and should only happen if the public interface is misused.

dtasev committed:
Added test_figureerrorsmanager to cmake

DTasev commented:
The scripted plots issue should be fixed. 

The context menu will now be shown _only_ if there is a line with errors plotted, and only work on that line's errors.

I haven't removed the show/hide all buttons yet.

dtasev committed:
Specify MPL backend, re #25547

hsautessella commented:
Another issue is x-errors. If you plot something with x and y errors and toggle the errorbars only one set of the errorbars will toggle. We probably want a menu option to toggle x-errors as well if they're available.

A script for x and y errorbars to make your life easier:
```python 
x = np.linspace(0, 15000, 150)
plt.figure()
plt.title('My Title')
plt.errorbar(x, 2*x, yerr=5*np.sqrt(2*x), xerr=2*np.sqrt(2*x), label='MyLabel')
plt.legend()
plt.show()
```
Also if you now drag a workspace on top of that figure, the first figure is not included in the legend. It's not immediately clear why, but I've had a little dig around and it could be when we're converting the Axes to a MantidAxes i.e. `MantidAxes.from_mpl_axes` the line being passed somehow loses its label and defaults to `_no_legend_`.

DTasev commented:
Should x-errors be a separate entry in the context menu, making it show 
```
Errors for Y -> all lines with errors on Y
Errors for X -> all lines with errors on X
```
Can lines have just the X error without Y errors?

hsautessella commented:
I think a separate entry would be good, but definitely get a second opinion. I don't see why a curve couldn't have only its x-errors showing, though admittedly it will probably be a rare-use case.

dtasev committed:
Merge branch 'master' into 25547_toggle_errorbars_on_plot

dtasev committed:
Fix toggling Y errors when X errors are present

Removes the `toggle` button.

The Show/Hide all and each line are now in an `Y Error Bars` context
menu.

Re #25547

dtasev committed:
Y errors can now be toggled even with X errors present

No operations are performed on the X errors. Showing/hiding errors only
works for Y errors.
`Toggle errors` button has been removed
Y errors options are now shown in a submenu
Test updated, and new tests added for scripted plots with X error
Ignore Flake8 import warning in test file

Re #25547

This commit squishes 11 debug commits, messages:

Added debug messages to check state on buildservers
Fix flake8 warning by changing var name
Fix ScriptedFigureErrors test failing, re #25547
Manually close axes to hopefully delete plots
Clear/close the figure and axes
Remove setUp/tearDown, use decorator
Manually clear lines from plot
Capture exceptions in decorator, re #25547
Fail test on exception
Reraise exception on exception
Add setUp/tearDown back

DTasev commented:
After an offline discussion with @martyngigg it was decided that the all the Y error related options should be in a submenu, now called `Y Error Bars`. 

Additionally, X errors are not shown/hidden when an action is performed. This could be a future extension, if required

DTasev commented:
Is there another way I am supposed to clear plots between tests? [This way](https://github.com/mantidproject/mantid/pull/25617/files#diff-9635e88c1c41bcb30b4d61224b16022eR50) seems to leave lines hanging around which causes all the test failures

dtasev committed:
Works correctly for plots with caps

Removed `test_add_errorbar_for` as `add_errorbar_for` is no longer part
of the public API.

Added test for get_data_lines

Re #25547

dtasev committed:
Add separator between show/hide all and individual lines

Re #25547

dtasev committed:
Merge branch 'master' into 25547_toggle_errorbars_on_plot

martyngigg commented on Framework/PythonInterface/mantid/plots/__init__.py:
```diff
@@ -763,6 +775,48 @@ def tricontourf(self, *args, **kwargs):
+        """
+        In the case where there are no arguments to the legend function,
+        but errors are present on the plot, we make our own legend order
+        to prevent MPL always appending errors on the bottom of the legend.
```
Could you help me understand the scenario where MPL appends the errors on the bottom of the legend? I haven't seen that before.

DTasev replied:
It seems that is the possibly unintended, but default behaviour when you have a mixture of lines that do not have `ErrorbarContainer`s associated with them, and lines that do. It happens because the `ax.containers` are added last [here](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/legend.py#L1212), which makes lines with errors always appear below lines without errors.

martyngigg reviewed (changes requested):
Code changes look good. A nice set of tests.

First issue I found when testing was when I loaded a workspace, plotted a spectrum *without* errors and then clicked "Hide all errors" and I got the image below:

![Screenshot from 2019-05-13 10-18-22](https://user-images.githubusercontent.com/733773/57609926-bfb67d80-7567-11e9-802d-de7424448973.png)

Another one was if I do a plot without errors then select show errors I get a set of error bars where the caps are different colours to the error bars:

![Screenshot from 2019-05-13 10-23-08](https://user-images.githubusercontent.com/733773/57610353-8d595000-7568-11e9-988a-f0838e5917b8.png)

dtasev committed:
Feedback and fixes from comments

Re #25547

- Removes duplicated import from plotfunctions.py
- Fixes `hide all error bars` adding error bar for lines that don't have
one. Now it just skips doing anything for lines without errors, while
still hiding the error bar on ones that do have it.
- Error caps should now inherit the color from the old line.
- `figureerrorsmanager.FigureErrorsManager.toggle_all_error_bars`'s
`make_visible` parameter is no longer a kwarg.
- Adjusted tests to change in API, and added new tests for
`make_visible` being False/True

DTasev commented:
Putting this to `In Progress` as I've found a couple of issues:
- Saving/loading project does not keep the caps lines' color. It also seems to make the references in the `ax.containers` wrong.
- `ax.creation_args` duplicates its contents on load -> I think when loaded back in and the correct `creation_args.function` is called, it adds another entry in the creation args.

Commit:
Fixes loading/saving for one/two consecutive saves.

Re #25547

After a 2-3 loads/saves of the same plots, it seems the lines get
reordered into actual "axes.errorbar" order, meaning ax.lines looks like
[data, cap1, cap2...], instead of [data, data2, data3, cap1, cap2, <rest
of caps>]. Further investigation needed as to why this happens. To
replicate plot 3-5 spectra, save project, load project, repeat until the
lines are incorrect.

TODO: use ERRORS_VISIBLE key when applying the lines style back in
plotloader

Commit:
Found a way to replicate issue:

Make Plot -> save -> load (while plot still open) -> lines are changed

It seems the figuremanager.replaceHandle is the source of the line
reordering. TODO: Investigate further!

Re #25547

mantid-builder commented:
@DTasev the build is failing, can you investigate.
<!-- {"problem_type": "failing"} -->

Commit:
Merge branch 'master' into 25547_toggle_errorbars_on_plot

TTitcombe committed:
Fix merge conflict in plotfunctions

Fix flake8 missing newline at end of plotsloader.py

TTitcombe committed:
On home button click, set ylim to cap bounds after default behaviour

re #25547

TTitcombe committed:
Fix merge conflict in plotfunctionsTest

TTitcombe committed:
Fix test indentation level in plotfunctionsTest

Incorrect indetation from bad merge

TTitcombe committed:
Fix uncaught KeyError for POST_CREATION_ARGS

Check if post creation args exist first

Only scale to error bars if post creation args have errors visible

TTitcombe committed:
When a project is loaded, plots keep errorbar toggling

Set default POST_CREATION_ARGS in plotssaver, so all plots can be saved

regardless or whether they have been changed post creation or not

re #25547

TTitcombe commented:
@DTasev I've fixed a couple of the bugs, however there are a few things that I think will need your input before this can continue, as I'm not familiar enough with the code: 
* there are a few commented out pieces of code you've introduced in previous commits. I'm not sure if these should be removed
* A few bugs were due to POST_CREATION_ARGS being assumed to exist. I've set default args at places where the errors were raised, however a better solution may be to set default POST_CREATION_ARGS with `None` values when creation args are created.
* There are a few merge conflicts I'm unsure how to deal with

Finally, I'm not convinced the issue of saving/loading mixing up the order of lines is an actual issue. Your instructions for replicating are:
* plot
* save project
* load project (while the plot is open)

When I closed the plot before re-opening, I saved/loaded several times without problem. I think the issue is when you have two plots open pointing to the same data. If that is the case, I'm not sure this would be a particularly important issue to fix.

DTasev commented on qt/python/mantidqt/project/plotsloader.py:
```diff
@@ -57,20 +58,32 @@ def make_fig(self, plot_dict, create_plot=True):
         # Populate workspace names
-        workspace_name = creation_args[0][0].pop('workspaces')
-        workspace = ADS.retrieve(workspace_name)
+        # workspace_name = creation_args[0][0].pop('workspaces')
```
These 3 lines can be removed I think, the code below doesn't use it anymore

DTasev commented on qt/python/mantidqt/project/plotsloader.py:
```diff
@@ -57,20 +58,32 @@ def make_fig(self, plot_dict, create_plot=True):
-        self.plot_func(workspace, ax, fig, creation_args[0][0])
-
-        # If an overplot is necessary plot onto the same figure
-        self.plot_extra_lines(creation_args=creation_args, ax=ax)
```
I don't think I've removed the implementation of this function, it might not be used anymore, and this functionality is merged into the `for cargs in creation_args[0]:` above

TTitcombe replied:
I've removed plot_extra_lines and the other comments

DTasev commented on qt/python/mantidqt/project/plotssaver.py:
```diff
@@ -64,7 +65,7 @@ def get_dict_from_fig(self, fig):
     @staticmethod
     def get_dict_for_axes_colorbar(ax):
-        image = None
+        # image = None
```
This was commented out as it was unused

DTasev commented on Framework/PythonInterface/test/python/mantid/plots/plotfunctionsTest.py:
```diff
@@ -139,7 +146,11 @@ def _do_update_colorplot_datalimits(self, color_func):
         funcs.update_colorplot_datalimits(ax, mesh)
         self.assertAlmostEqual(10.0, ax.get_xlim()[0])
+        self.assertAlmostEqual(30.0, ax.get_xlim()[1])
+        # self.assertAlmostEqual(4.0, ax.get_ylim()[0])
```
I think the lines above are the new values

DTasev commented on qt/applications/workbench/workbench/plotting/figureerrorsmanager.py:
```diff
@@ -0,0 +1,343 @@
+                if MantidAxKwargs.POST_CREATION_ARGS in creation_args[index]:
+                    creation_args[index][MantidAxKwargs.POST_CREATION_ARGS][MantidAxKwargs.ERRORS_VISIBLE] = new_state
+                else:
+                    creation_args[index][MantidAxKwargs] = {MantidAxKwargs.ERRORS_VISIBLE: new_state}
```
Is this missing a `.POST_CREATION_ARGS`?

TTitcombe replied:
Yes, good spot

dtasev reviewed (commented):
>there are a few commented out pieces of code you've introduced in previous commits. I'm not sure if these should be removed

Hopefully I got all of them in the review

>A few bugs were due to POST_CREATION_ARGS being assumed to exist. I've set default args at places where the errors were raised, however a better solution may be to set default POST_CREATION_ARGS with None values when creation args are created.

This is sensible, but I think we still need to keep the check if `POST_CREATION_ARGS` exists otherwise plots saved in 4.0 might error after this PR.

I agree the saving/loading is lower priority than I previously thought

Commit:
Merge branch 'master' into 25547_toggle_errorbars_on_plot

TTitcombe committed:
Remove commented out code

Fix flake8 duplicate import error

Fix issues arising from bad merge

Readd missing import

MantidAxType and MantidAxKwargs were removed from plotfunctions during merge

include default post_creation_args in loader dicts in failing tests

Test that existing post_creation_args are not overwritten on save

Move duplicate xlim check to within if/else statement

Bad merge had moved it outside of statement

Check needs to be in statement to account for matplotlib version differences

TTitcombe commented:
@martyngigg @hsautessella @DTasev 
This is ready for another review

martyngigg commented:
Thanks Tom!

DTasev commented on qt/applications/workbench/workbench/plotting/figuremanager.py:
```diff
@@ -283,6 +285,30 @@ def destroy(self, *args):
+                for container in ax.containers:
+                    caps = container[1]
+                    if caps != ():
+                        min_bounds.extend(caps[0].get_ydata())
```
I think these may be `numpy.ndarray`. If so they should have `.min()`/`.max()` methods

DTasev commented on qt/applications/workbench/workbench/plotting/figuremanager.py:
```diff
@@ -283,6 +285,30 @@ def destroy(self, *args):
+                        max_bounds.extend(caps[1].get_ydata())
+        if min_bounds and max_bounds:
+            # Set ylim if we have caps
+            ax.set_ylim((min(min_bounds), max(max_bounds)))
```
@TTitcombe I think MPL sets some margins on the limits, as there's always some space between the data and the borders of the plot. Currently, after showing the errors and going to the "home" view the data seems to end exactly at the border of the plot. 

Do you think it'd be possible to the margins back while errors are shown? I found [this](https://stackoverflow.com/a/42804403/2823526) which mentions this method `ax.set_ymargin(0.1)`, alternatively also gives this `ymax = data.max()+0.4*(data.max()-data.min())` for a 40% margin (I think about 10%-15% should be enough). Another possible route is to check how MPL does it, but that would involve some digging in the source

TTitcombe replied:
Good idea. I've calculated the margin manually by adding padding by 5% of the range of the data to the top and bottom (`set_ymargin` didn't work)

TTitcombe committed:
Add margin to errorbar plots on home button press

Include line points when determining plot bounds

Correct home button behaviour on errorbar plot with hidden errors

Add figuremanager tests for finding container bounds

TTitcombe commented:
@DTasev I've included your comments, and I've fixed a case where a spectrum with errors plot was being scaled incorrectly when the errors had been hidden.

I've added a couple of tests to `test_figuremanager` to check that the correct plot bounds are being selected, however the test suite takes ~3s to run for only 4 tests. I think the issue is related to mocking out QAppThreadCall everytime. Is it necessary to include this in the tests?

If we want to expand the functionality of figuremanager, we'll have to figure out a way to test the functionality without including the thread-mocking overhead

TTitcombe committed:
Remove double numpy import in figuremanager

Fixes flake8 failure

TTitcombe committed:
Remove None values from list before finding max

max and min raises errors on python3 if list includes None values

TTitcombe committed:
Remove whitespace from blank line

fix flake8 error

martyngigg commented on Framework/PythonInterface/mantid/plots/utility.py:
```diff
@@ -18,3 +18,22 @@ class MantidAxType(Enum):
+
+
+class MantidAxKwargs(object):
+    POST_CREATION_ARGS = "post_creation_args"
```
Is this in fact used in the `mantid.plots` package? Should it belong in the `figureerrorsmanager`?

TTitcombe replied:
It's used in `workbench.plotting`, `mantid.plots`, and `mantidqt.project`, so I'm not sure where the best place for it would be

martyngigg commented:
@TTitcombe Do you think you would have time to look over the final few things here? If not I could get Harry to take a look.

TTitcombe commented:
@martyngigg I can look at this

TTitcombe committed:
Remove leftover debugging print statements

re #25547

TTitcombe committed:
Rename MantidAxKwargs to MantidAxPostCreationArgs

Add class docstring to MantidAxPostCreationArgs

Turn MantidAxPostCreationArgs into an Enum class

re #25547

hsautessella commented:
I'm getting errors on this whenever I try to show errorbars.
To recreate:
- Run:
```python
ws2d_1spec = CreateWorkspace(DataX=[10, 20, 30],
                                    DataY=[2, 3],
                                    DataE=[1, 2],
                                    NSpec=1,
                                    Distribution=True,
                                    UnitX='Wavelength',
                                    VerticalAxisUnit='DeltaE',
                                    VerticalAxisValues=[4])
```
- `Right-click on workspace -> Plot -> Spectrum`. 
- `Right-click on plot -> Y Error Bars -> Show All Errors`. Gives me the error:

```
Traceback (most recent call last):
  File "c:\users\ejo73213\mantid\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 337, in _update_plot_after
    func(*args, **kwargs)
  File "c:\users\ejo73213\mantid\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 271, in toggle_all_error_bars
    self._add_errorbar_for(index, data_line, make_visible)
  File "c:\users\ejo73213\mantid\mantid\qt\applications\workbench\workbench\plotting\figureerrorsmanager.py", line 144, in _add_errorbar_for
    specNum = ax.creation_args[creation_args_index]["specNum"]
KeyError: 'specNum'
```
Also toggling normalization seems to break when after toggling errorbars

martyngigg commented:
Are we lacking a developer on this now?

martyngigg commented:
@DTasev Any chance you would have some time to polish this off?

Commit:
Merge branch 'master' into 25547_toggle_errorbars_on_plot

DTasev commented:
I'll have a look at this after #25945 is merged - I believe the functionality from it can be reused here

martyngigg commented:
@DTasev #25945 has gone in now.

Commit:
Merge branch 'master' into 25547_toggle_errorbars_on_plot

Commit:
Error bars on plots can now be toggled from the context menu

- Now reuses work from the plot config window, to toggle the errorbars in a similar manner. This significantly reduced the amount of new code
- The default color of the error bars is now the color of the line, not the default color from rcParams

Re #25547

Commit:
Revert changes to plotfunctions, as they are not necessary anymore

- Revert plotfunctions.errorbar changes
- Remove custom legend override

Commit:
FigureErrorsManager now uses logic from the plot properties

Uses static functions from CurveProperties and CurvesWidgetPresenter to toggle error bars on lines.
Tests adapted to new functionality

Re #25547

Commit:
Removes FigureErrorsManager usage from plotsloader

- Makes plotsloader more verbose about why a plot wasn't loaded
- Makes plotssaver more verbose about why the JSON file couldn't be written to
- Some formatting changes

Commit:
Remove comments, re #25547

Commit:
Fix tests and flake 8, re #25547

Remove blank line for flake8
Fix tests after removal of legend override
Remove MPL_NOLEGEND and errors_visible parameter

DTasev commented on qt/python/mantidqt/project/test/test_plotssaver.py:
```diff
@@ -42,34 +42,35 @@ def setUp(self):
-                                                                      u'position': u'Left',
-                                                                      u'visible': True},
-                                                 u'yAxisScale': u'linear', u'yLim': (0.0, 1.0)},
+                                            u'properties': {u'axisOn': True, u'bounds': (0.0, 0.0, 0.0, 0.0),
```
This is just from formatting - I haven't changed any of the values

hsautessella commented on qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/presenter.py:
```diff
@@ -56,12 +56,16 @@ def apply_properties(self):
         curve = self.get_selected_curve()
         # Set the curve's new name in the names dict and combo box
         self.set_new_curve_name_in_dict_and_combo_box(curve, view_props.label)
+        self.toggle_errors(self.get_selected_ax(), curve, view_props)
```
It looks like we're calling `legend` and calling `relim` twice here. Once in `replot_selected_curve` and then again in `toggle_errorbars`. We probably want to take `relim` and `legend` out of these functions and call them at the end of `apply_properties` and at the end of `FigureErrorsManager.toggle_errorbars_for`?

DTasev replied:
I've refactored the code that decides how to replot the curve into another function, which is now reused from both the `CurvesTabWidgetPresenter` and `FigureErrorsManager`. `apply_properties` now uses that. Additionally the relim/scale/legend is now in `update_limits_and_legend` called from both classes separately

hsautessella commented on qt/applications/workbench/workbench/plotting/figureerrorsmanager.py:
```diff
@@ -0,0 +1,130 @@
+            try:
+                new_curve = ax.replot_artist(curve, errorbars=True, **plot_kwargs)
+            except ValueError:  # ValueError raised if Artist not tracked by Axes
+                # this might not have a creation_args
```
```suggestion
                # this might not have creation_args
```

hsautessella commented on qt/applications/workbench/workbench/plotting/figureerrorsmanager.py:
```diff
@@ -0,0 +1,130 @@
+    @staticmethod
+    def toggle_error_bars_for(ax, curve, make_visible=None):
+        curve_props = CurveProperties.from_curve(curve)
+        plot_kwargs = {key: value for key, value in curve_props.items() if key not in ['hide', 'hide_errors']}
```
I think because we're not converting the `hide` argument into the `visible` Matplotlib kwarg if you set a curve to hidden in the plot options, and then toggle errorbars, the curve will no longer be hidden. I don't know if this is desired behaviour or not, but if we do want to keep the curve hidden in this case, we just need this line to be changed to
```python
plot_kwargs = curve_props.get_plot_kwargs()
```

DTasev replied:
I am not sure about the behaviour, but this lets me reuse the same code as the presenter, which I'm all for

martyngigg commented:
I'll leave this one to Harry.

DTasev commented on qt/applications/workbench/workbench/plotting/figuremanager.py:
```diff
@@ -186,6 +188,7 @@ def __init__(self, canvas, num):
         if self.toolbar is not None:
             self.window.addToolBar(self.toolbar)
             self.toolbar.message.connect(self.statusbar_label.setText)
+            # self.toolbar.home_clicked.connect(self.on_home_clicked)
```
leftover code

DTasev commented on qt/applications/workbench/workbench/plotting/test/test_figureerrorsmanager.py:
```diff
@@ -0,0 +1,210 @@
+
+import matplotlib
+
+matplotlib.use("AGG")
```
This is a bit sneaky but `MantidAxes` is used by the `type` specification [here](https://github.com/mantidproject/mantid/pull/25617/files/7a82ace5caf98a3a1ce5f3db7c7f1b5fe4e247ab#diff-9635e88c1c41bcb30b4d61224b16022eR129), as well as the projection registration. I think it's worth the comment nonetheless

DTasev replied:
Also I didn't know I could just stick in a `# noqa` and the lines below will be unaffected - very nice!

DTasev replied:
Nevermind! Flake8 didn't recognise the type usage :(

DTasev commented on qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/__init__.py:
```diff
@@ -125,14 +125,24 @@ def __getattr__(self, item):
         return kwargs
 
+    @classmethod
+    def get_mpl_plot_kwargs_from_curve(cls, curve):
```
:+1: Good spot, it's not used anywhere anymore!

Commit:
Remove duplicated code from plots/__init__

Re #25547

DTasev commented on qt/applications/workbench/workbench/plotting/figureerrorsmanager.py:
```diff
@@ -0,0 +1,130 @@
+                                  partial(self._update_plot_after, self._toggle_all_errors, ax, make_visible=False))
+        parent_menu.addMenu(error_bars_menu)
+
+        self.active_lines = CurvesTabWidgetPresenter.get_active_lines(
```
I agree it doesn't feel right, and I have opened a maintenance task to refactor it: https://github.com/mantidproject/mantid/issues/26330

Commit:
Apply feedback from comments

- Refactored the CurveTabsWidgetPreseter replot_curve functions: now has a common `replot_curve` that does the replotting of all curves. Internally it delegates to `_replot_mpl_curve` if the curve is not on a MantidAxes. This `replot_curve` function is now used within FigureErrorsManager as well reducing code duplication.
- Relimiting, autoscaling and updating the legend are now wrapped in a `update_limits_and_legend` function, used from both CurveTabsWidgetPreseter and FigureErrorsManager.
- Made some of the CurveTabsWidgetPreseter replotting functions private.
- Deleted commented code from figuremanager.py.

Re #25547

DTasev commented:
@hsautessella Ready for another review

DTasev commented:
@martyngigg @NickDraper The issue #25547 for this has been re-targeted for 4.2, shall we still aim to get this in 4.1 in this case?

martyngigg commented:
@DTasev Yes, please pull the issue back

Commit:
Add noqa for flake8 unused MantidAxes import, re #25547

Commit:
Keep errorbar state after changing Normalisation

Code from @hsautessella

Bug not related to this PR, but fixed here anyway

hsautessella commented on qt/python/mantidqt/widgets/plotconfigdialog/curvestabwidget/presenter.py:
```diff
@@ -246,15 +259,25 @@ def _populate_select_curve_combo_box(self):
+        return ax.get_lines() + CurvesTabWidgetPresenter.get_errorbars_from_ax(ax)
+
+    @staticmethod
+    def get_active_lines(selected_ax, errorbars):
```
I think you may have forgotten to take this function out

Commit:
Remove unused function, re #25547

hsautessella commented:
This looks good to me 👍

DTasev commented:
There's some leftover debug code in `figuremanager.py`, I'm taking a look

NickDraper reviewed (changes requested):
Release notes are in the wrong place and a resurrecting a dead file that had otherwise been removed.  Please move the line to the workbench.rst.

No need to wait for the builds after this, just give me a nudge and I'll merge it.

NickDraper commented:
I made the release note changes
